lib/eco/treelexer/lexer.py

```diff
@@ -49,25 +56,50 @@ def match_question(self, pattern, text):
         return self.match(pattern[2:], text)
 
     def match_star(self, pattern, text):
-        if self.match_one(pattern, text):
-            # matched one, continue matching *
-            return self.match(pattern, text[1:])
-        # couldn't match star, remove star regex and continue
-        return self.match(pattern[2:], text)
+        while True:
+            if not self.match(pattern.c, text):
+                break
+        return True
+
+    def match_plus(self, pattern, text):
+        if not self.match(pattern.c, text):
+            # we have to at least match one
+            return False
+        while True:
+            if not self.match(pattern.c, text):
+                break
+        return True
+
+    def match_list(self, pattern, text):
+        for p in pattern:
+            if not self.match(p, text):
+                return False
+        return True
+
+    def match_or(self, pattern, text):
+        tmp = self.pos
```

Comment:
Ah, I wondered when doing `self.pos` was going to bite you ;)

I'm not sure it's worth changing things now, but I *much* prefer passing and returning `pos` as a parameter, so that you can "undo" it without having to do global mutation. [This is how things like lrtable work FWIW.]

Reply:
Indeed. I didn't remember that `or` introduces backtracking. You'll see in a later commit that I need to store even more state, so at least it prepared me for that. ;-)